### PR TITLE
Add type hints to `sygnal/utils.py`

### PR DESCRIPTION
**Pull request by H-Shay**
As the title states.

### CODE REVIEW

File: sygnal/utils.py [hunk 1]
```diff
@@ -65,7 +71,7 @@ def glob_to_regex(glob):
     return re.compile(r"\A" + res + r"\Z", re.IGNORECASE)
 
 
-def _reject_invalid_json(val):
+def _reject_invalid_json(val: Any) -> None:
```
Comment (DMRobertson) [inline sygnal/utils.py:74]: Could also use `NoReturn` here, but I don't think there's any practical benefit to doing so.

File: sygnal/utils.py [hunk 2]
```diff
@@ -37,17 +41,19 @@ async def twisted_sleep(delay, twisted_reactor):
...
 class NotificationLoggerAdapter(LoggerAdapter):
-    def process(self, msg, kwargs):
+    def process(
+        self, msg: str, kwargs: MutableMapping[str, Any]
+    ) -> Tuple[str, MutableMapping[str, Any]]:
         return f"[{self.extra['request_id']}] {msg}", kwargs
 
 
-def glob_to_regex(glob):
+def glob_to_regex(glob: str) -> Union[Pattern, Pattern[str]]:
```
Comment (DMRobertson) [inline sygnal/utils.py:50]: I don't understand the return type here---why can't we just say it returns a `Pattern[str]` in all cases?

File: sygnal/utils.py [hunk 3]
```diff
@@ -15,11 +15,15 @@
...
+from typing import TYPE_CHECKING, Any, MutableMapping, Pattern, Tuple, Union
 
 from twisted.internet.defer import Deferred
 
+if TYPE_CHECKING:
+    from sygnal.sygnal import SygnalReactor
 
-async def twisted_sleep(delay, twisted_reactor):
+
+async def twisted_sleep(delay: float, twisted_reactor: "SygnalReactor") -> None:
```
Comment (DMRobertson) [inline sygnal/utils.py:26]: This is fine as it is, but I wanted to mention a couple of alternatives here. One be to write `twisted_reactor: IReactorTime` and only import `IReactorTime`; a second would be to move `SygnalReactor` into a new file, say `sygnal/types.py`.

File: sygnal/utils.py (hunk 2 above)
Comment (H-Shay) [reply sygnal/utils.py:50]: I nicked this from somewhere and now I can't remember where, but I think you are correct we can use the narrower case. I will change this.

File: sygnal/utils.py (hunk 3 above)
Comment (H-Shay) [reply sygnal/utils.py:26]: Great, I appreciate knowing alternative solutions, it really helps to have more tools in my arsenal.

**Review by DMRobertson — APPROVED** [review]
I'm happy if mypy is!

**Pull request merged**